fastavro/_write.pyx

```diff
@@ -521,21 +543,43 @@ cpdef write_union(bytearray fo, datum, schema):
     write_data(fo, datum, schema[index])
 
 
-cpdef write_record(bytearray fo, dict datum, dict schema):
+cpdef write_record(bytearray fo, object datum, dict schema):
     """A record is encoded by encoding the values of its fields in the order
     that they are declared. In other words, a record is encoded as just the
     concatenation of the encodings of its fields.  Field values are encoded per
     their schema."""
     cdef list fields
     cdef dict field
+    cdef dict d_datum = None
     fields = schema['fields']
-    for field in fields:
-        name = field['name']
-        if name not in datum and 'default' not in field and\
-                'null' not in field['type']:
-            raise ValueError('no value and no default for %s' % name)
-        write_data(fo, datum.get(
-            name, field.get('default')), field['type'])
+    try:
```

Comment:
If we wrote this `try` block like the following I don't think there would be a need for the test case that checks the stack trace (as I think it should be the same):

```python
    try:
        d_datum = <dict?>(datum)
    except TypeError:
        # Slower, general-purpose code where datum is something besides a dict,
        # e.g. a collections.OrderedDict or collections.defaultdict.
        for field in fields:
            name = field['name']
            if name not in datum and 'default' not in field and \
                    'null' not in field['type']:
                raise ValueError('no value and no default for %s' % name)
            write_data(fo, datum.get(
                name, field.get('default')), field['type'])
    else:
        # Faster, special-purpose code where datum is a Python dict.
        for field in fields:
            name = field['name']
            if name not in d_datum and 'default' not in field and \
                    'null' not in field['type']:
                raise ValueError('no value and no default for %s' % name)
            write_data(fo, d_datum.get(
                name, field.get('default')), field['type'])
```

At least it seemed the same when I did some quick testing...

Reply:
Oh yeah, this looks like a good situation for `try/except/else` -- I didn't think of that. I'll try that out and make sure the generated code looks reasonable, performance wise. In theory, straight-line execution is fastest so it'd be nice not to have the `except` in the middle, but I think that's likely to be negligible impact here.

Reply:
I went ahead and cleaned up the library code as you suggested. I'd like to keep the stack trace tests, even though they are kind of ugly and "white box", because they do help protect against future regression bugs where `dict` and other types end up using the same, slower code.

Reply:
(BTW, the change you suggested did involve updating the stack trace tests, in case that was not clear from the above comment. They are still quite similar to before, only the specific stack trace checks changed a bit.)